zkevm-circuits/src/evm_circuit/execution/blockhash.rs

```diff
@@ -0,0 +1,201 @@
+                BlockContextFieldTag::BlockHash.expr(),
+                Some(from_bytes::expr(&block_number.cells)),
+                block_hash.expr(),
+            );
+        });
+        cb.condition(not::expr(block_lt.expr() * diff_lt.expr()), |cb| {
+            cb.require_zero("invalid range", block_hash.expr());
+        });
+
+        cb.stack_push(block_hash.expr());
+
+        let step_state_transition = StepStateTransition {
+            rw_counter: Delta(2.expr()),
+            program_counter: Delta(1.expr()),
+            gas_left: Delta(-OpcodeId::BLOCKHASH.constant_gas_cost().expr()),
+            ..Default::default()
+        };
+
+        let opcode = cb.query_cell();
+        let same_context = SameContextGadget::construct(cb, opcode, step_state_transition);
+        Self {
+            same_context,
+            block_number,
+            current_block_number,
+            block_hash,
+            block_lt,
+            diff_lt,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _: &Transaction,
+        _: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let block_number = block.rws[step.rw_indices[0]].stack_value();
+        self.block_number.assign(
+            region,
+            offset,
+            Some(
+                block_number.to_le_bytes()[..N_BYTES_U64]
+                    .try_into()
+                    .unwrap(),
+            ),
+        )?;
+        let block_number: F = block_number.to_scalar().unwrap();
+
+        let current_block_number = block.context.number;
+        self.current_block_number
+            .assign(region, offset, current_block_number.to_scalar())?;
+
+        self.block_hash.assign(
+            region,
+            offset,
+            Some(block.rws[step.rw_indices[1]].stack_value().to_le_bytes()),
+        )?;
+
+        self.block_lt.assign(
+            region,
+            offset,
+            block_number,
+            current_block_number.to_scalar().unwrap(),
+        )?;
+
+        self.diff_lt.assign(
+            region,
+            offset,
+            current_block_number.to_scalar().unwrap(),
+            block_number + F::from(256),
+        )?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{
+        evm_circuit::witness::block_convert,
+        test_util::{test_circuits_using_witness_block, BytecodeTestConfig},
+    };
+    use bus_mapping::mock::BlockData;
+    use eth_types::{bytecode, geth_types::GethData, Bytecode, U256};
+    use mock::test_ctx::{helpers::*, TestContext};
+
+    fn test_ok(block_number: usize, current_block_number: u64) {
+        let mut code = Bytecode::default();
+        code.append(&bytecode! {
+            PUSH32(block_number)
+            BLOCKHASH
+            STOP
+        });
```

Comment:
Just easier to create the code directly with the macro instead of using the `append` fn.

Reply:
Fixed. Thanks!